### ISSUE TITLE
ReferenceArrayInput: too many requests to getMany

### ISSUE THREAD

**Issue by joanllodrariera** [marmelab/react-admin#8942]
**Is your feature request related to a problem? Please describe.**

In the `ReferenceArrayInput` as indicated in the documentation the component is using the `getMany` to get the data of the ids that the response may contain. As per the documentation:

![image](https://github.com/marmelab/react-admin/assets/87600529/8e20183b-bb43-4744-a098-09c95177f386)

The problem from my point of view is that the component is also executing this `getMany` for new records. Meaning that if I select a new record obtained from the `getList` method once I've selected it, it executes a `getMany` for this record. Which doesn't make sense as the new record information was already returned from the `getList` method so it should be available.

**Describe the solution you'd like**

The solution should avoid calls to the `getMany` method for records already obtained from the `getList`. This will improve the performance of the component.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by slax57** [comment]
Hi,

Thank you for your suggestion, and sorry for the late answer.

It's true that we could in theory pre-populate the `getMany` cache with the result of `getList`, however this seems hard to do in practice, at least with react-admin's current architecture.
Indeed the `getMany` cache is handled by react-query, and is linked to the query key which is built like so:
```ts
[
    resource,
    'getMany',
    {
        ids: !ids || ids.length === 0 ? [] : ids.map(id => String(id)),
        meta,
    },
],
```

So the cache corresponding to the query `['posts', 'getMany', {ids: [1, 2, 3]}]` is not the same as the cache corresponding to the query `['posts', 'getMany', {ids: [1, 2, 3, 4]}]`.
This means we cannot simply pre-populate the cache from the result of `getList`, and it would require more complex logic to populate it on click when a user selects a choice from a `SelectInput` or an `AutocompleteInput`.

Besides, since by default react-query uses a `staleTime` of `0`, an API call to `getMany` would still be made.

TLDR: I believe this would be hard to do in practice, at least without introducing breaking changes to the core.

@djhi @fzaninotto what's your take on this?

**Comment by fzaninotto** [comment]
I agree, this would imply populating *all* the possible combinations if ids in the getMany cache. For a getList with 3 results (a, b, c), this means 13 cache keys to populate:

- a
- b
- c
- a, b
- a, c
- b, a
- c, a
- b, c
- c, b
- a, b, c
- a, c, b
- b, c, a
- b, a, c

That doesn't scale at all. 

So we won't implement that feature. 

Thanks for the suggestion anyway!
